### DISCUSSION TITLE
[For Brown Community] Setting up HSSM on Oscar

### DISCUSSION THREAD

**Discussion by digicosmos86** [lnccbrown/HSSM#440] [Announcements]
This is a tutorial on how to get HSSM to work on Oscar, Brown University's HPC cluster. This may or may not work with other HPC environments, depending on the setup.

You can choose to work with HSSM either in a `venv` or a `conda` environment. The former option, if linked correctly with the Intel MKL library on Oscar, can lead to a significant performance boost on Oscar.

## Option 1: Using HSSM in `venv` (Faster)

In general, we recommend that we work with environment created with `venv` on Oscar and install packages with `pip`. However, the set up for PyMC is slightly more involved because we need to tell PyMC where it can find the BLAS library to accelerate its computation. Fortunately, we only need to do this once, unless the path to the library changes. When that happens, we just need to run this process again.

### Create a `.pytensorrc` file with path to a BLAS library

The BLAS library that seems to work well with `pytensor` is the Intel one MKL library. We need to load the `intel-oneapi-mkl` module and tell `pymc` how to link to that library:

```
module purge
unset LD_LIBRARY_PATH
module load intel-oneapi-mkl
echo $LD_LIBRARY_PATH
```

You will see a few paths printed out separated with colons. Copy the path that ends in `intel64` after running the above command. Then, create a `.pytensorrc` file under your home directory `~`, and add the following content:

```
[blas]
ldflags=-Lpath_just_copied -Wl,--no-as-needed -lmkl_gf_lp64 -Wl,--start-group -lmkl_gnu_thread -lmkl_core  -Wl,--end-group -fopenmp  -ldl -lpthread -lm
```

Replace `path_just_copied` with what you have just copied. Be sure to keep the `-L` before the path. This will ensure that `pymc` finds the lapack library. You only need to do this once unless the LAPACK module gets changed. The file will look like this:

```
[blas]
ldflags=-L/oscar/rt/9.2/software/0.20-generic/0.20.1/opt/spack/linux-rhel9-x86_64_v3/gcc-11.3.1/intel-oneapi-mkl-2023.1.0-xbcd2g3myicmoznxp4pajepy7y2xzsvh/mkl/2023.1.0/lib/intel64 -Wl,--no-as-needed -lmkl_gf_lp64 -Wl,--start-group -lmkl_gnu_thread -lmkl_core  -Wl,--end-group -fopenmp  -ldl -lpthread -lm
```

### Load modules and create environment

We need the `python` module to create the virtual environment and activate it. You can do this in the login node.

```sh
module load python intel-oneapi-mkl
python -m venv .venv
source .venv/bin/activate
```

### Install HSSM

Once the environment is loaded, we can install `hssm` with `pip`:

```bash
pip install hssm
```

Or you can install the `dev` version directly from github:
```bash
pip install git+https://github.com/lnccbrown/HSSM.git
```

The installation is complete. Please ensure that there is no warning like this: `WARNING (pytensor.tensor.blas): Using NumPy C-API based implementation for BLAS functions.` when `hssm` is imported.

### Optional: Install JAX with GPU support

If you need JAX with GPU support, please perform the installation in an interactive GPU node:

```sh
interact -q gpu -g 1 -f ampere -m 20g -n 4
```

Once the node is up, run the following commands to set up the virtual environment. Note that you also need to load the `cuda` and `cudnn` modules before proceeding:

```sh
module load python intel-oneapi-mkl cuda cudnn    # MKL is optional here
python -m venv .venv
source .venv/bin/activate
```

Then, install the GPU version of `JAX`:

```sh
pip install jax[cuda12]
```

Finally, you can choose to install the release or dev versions of HSSM.

## Option 2: Using HSSM in a conda environment:

The use of `conda` is discouraged on Oscar. It might be slow in resolving dependencies. If you have to use `conda`, we recommend using `mamba` in place of `conda` for considerable speed-up.

With the conda environment, you do **NOT** need to set up the `.pytensorrc` file. In fact, it's not a good idea to have that file in your home directory.

### Load modules and create environment

We need the `miniforge` module to create the virtual environment and activate it. It comes with `mamba` installed. You can also use the `anaconda` module, but you won't be able to use mamba with it.

```sh
module load miniforge
mamba create -n my_env python=3.11 #3.10 also works
mamba activate my_env
```

### Install HSSM

HSSM is directly available on `conda-forge` now, so you can directly install it:

```sh
mamba install -c conda-forge hssm
```

If you need the dev version of HSSM, then you need to install `pymc` from conda first, before installing the dev version from github with pip:

```sh
mamba install -c conda-forge pymc
pip install git+https://github.com/lnccbrown/HSSM.git
```

### Optional: Install JAX with GPU support

If you need JAX with GPU support, please perform the installation in an interactive GPU node:

```sh
interact -q gpu -g 1 -f ampere -m 20g -n 4
```

Once the node is up, run the following commands to set up the conda environment. Note that you also need to load the `cuda` and `cudnn` modules before proceeding:

```sh
module load miniforge cuda cudnn
mamba create -n my_env python=3.11 #3.10 also works
mamba activate my_env
```

Then, install the GPU version of `JAX`:

```sh
pip install jax[cuda12]
```

Finally, follow the instruction in the previous section to install either the release or the dev version of HSSM.

**Comment by igrahek** [comment]
Getting this error on Oscar: 

```bash
[igrahek@login009 ~]$ module load intel-one-mkl
Lmod has detected the following error:  The following module(s) are unknown: "intel-one-mkl"
Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "intel-one-mkl"
Also make sure that all modulefiles written in TCL start with the string #%Module
```

**Reply by digicosmos86** [reply]
Sorry the module should be `intel-oneapi-mkl`. I've updated the document

**Comment by mcfreund** [comment]
Hello, a clarification question about which modules to load when: In the GPU example, `netlib-lapack` is loaded. Should this be loaded anytime jax GPU backend is being used, e.g. when training LANs via `lanfactory`? (I was under impression it was for CPU computation, not GPU.)

**Reply by digicosmos86** [reply]
Hi Mike,

You are right. This does not affect the GPU backend, which relies on JAX for computation

**Reply by mcfreund** [reply]
OK thanks. JW, what's the motivation for loading it here? (& why not MKL instead?)

**Reply by digicosmos86** [reply]
You got me lol. I actually had to go through edit history to see why `netlib-lapack` was written here. It turns out that I was recommending `netlib-lapack` for both CPU and GPU options. Then I realized that MKL is a better option, so I updated the CPU section and forgot to update the GPU section. Will fix the inconsistency. Thanks for pointing this out @mcfreund!

**Comment by mcfreund** [comment]
HSSM install is failing due to unavailability of `absl-py` package. Development version did not fix, but installing `absl-py` did.

Install commands:
```
module purge
module load python intel-oneapi-mkl
uv venv
source .venv/bin/activate
uv pip install hssm
```

Running README example:

```
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/oscar/data/dbadre/mfreund/prep/src/ssm/tests/test_hssm.py", line 1, in <module>
    import hssm
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/hssm/__init__.py", line 14, in <module>
    from .config import ModelConfig
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/hssm/config.py", line 13, in <module>
    from .defaults import (
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/hssm/defaults.py", line 34, in <module>
    default_model_config: DefaultConfigs = {"ddm": get_default_model_config("ddm")}
                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/hssm/modelconfig/__init__.py", line 69, in get_default_model_config
    module = importlib.import_module(f".{module_name}", package="hssm.modelconfig")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/oscar/runtime/software/external/python/3.11.0/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/hssm/modelconfig/ddm_config.py", line 2, in <module>
    from ..likelihoods.analytical import ddm_bounds, ddm_params, logp_ddm
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/hssm/likelihoods/__init__.py", line 3, in <module>
    from .analytical import DDM, DDM_SDV, logp_ddm, logp_ddm_sdv
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/hssm/likelihoods/analytical.py", line 17, in <module>
    from ..distribution_utils.dist import make_distribution
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/hssm/distribution_utils/__init__.py", line 3, in <module>
    from .dist import (
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/hssm/distribution_utils/dist.py", line 29, in <module>
    from .onnx import (
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/hssm/distribution_utils/onnx.py", line 20, in <module>
    from .onnx_utils.onnx2jax import make_jax_func
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/hssm/distribution_utils/onnx_utils/onnx2jax.py", line 8, in <module>
    from jaxonnxruntime import call_onnx
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/jaxonnxruntime/call_onnx.py", line 22, in <module>
    from jaxonnxruntime import onnx_ops  # pylint: disable=unused-import
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/jaxonnxruntime/onnx_ops/__init__.py", line 22, in <module>
    from . import abs as abs
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/jaxonnxruntime/onnx_ops/abs.py", line 37, in <module>
    from jaxonnxruntime.core import onnx_node
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/jaxonnxruntime/core/onnx_node.py", line 31, in <module>
    from .onnx_utils import contain_subgraph
  File "/oscar/data/dbadre/mfreund/prep/.venv/lib/python3.11/site-packages/jaxonnxruntime/core/onnx_utils.py", line 20, in <module>
    from absl import logging
ModuleNotFoundError: No module named 'absl'

```

**Reply by digicosmos86** [reply]
Confirming this is an issue with running hssm on Oscar. It's weird because our `jaxonnxruntime` dependency explicitly depends on `absl-py` yet neither `uv` nor `pip` installs it. We need to explicitly add `absl-py` to the dependency of hssm to resolve this
